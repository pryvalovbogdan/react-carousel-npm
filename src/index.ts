import MyCounter from './components/App';

export { MyCounter };
