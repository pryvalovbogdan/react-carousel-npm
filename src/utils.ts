import React, { ReactNode } from 'react';

export const renderChildren = (children: React.ReactNode, props?: any) => {
  return React.Children.map<ReactNode, ReactNode>(children, child => {
    if (React.isValidElement(child)) {
      return React.cloneElement(child, props);
    }

    return child;
  });
};

export const renderChild = (child: React.ReactNode, props?: any) => {
  if (React.isValidElement(child)) {
    return React.cloneElement(child, props);
  }

  return child;
};
