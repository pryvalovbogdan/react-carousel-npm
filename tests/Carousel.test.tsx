import * as React from 'react';
import '@testing-library/jest-dom';
import { fireEvent, render } from '@testing-library/react';

import 'jest-canvas-mock';

import { Carousel, CarouselContextProvider } from '../src';

const mockCards = [
  { key: '1', title: 'Card 1' },
  { key: '2', title: 'Card 2' },
  { key: '3', title: 'Card 3' },
];

const Card = (props: any) => <div>{props.title}</div>;
describe('Carousel test render', () => {
  it('snapshot Carousel', () => {
    const { baseElement } = render(
      <Carousel i18n={'Cards'} cards={mockCards} defaultCardsCount={1}>
        <Card />
      </Carousel>,
    );

    expect(baseElement).toMatchSnapshot();
  });

  test('renders carousel container', () => {
    const { getByTestId } = render(
      <Carousel i18n={'Cards'} cards={mockCards} defaultCardsCount={1}>
        <Card />
      </Carousel>,
    );
    const carouselContainer = getByTestId('carousel-Cards');

    expect(carouselContainer).toBeInTheDocument();
  });

  it('handles next page click', () => {
    const { getByTestId } = render(
      <CarouselContextProvider>
        <Carousel i18n={'Cards'} cards={mockCards} defaultCardsCount={1}>
          <Card />
        </Carousel>
      </CarouselContextProvider>,
    );
    const nextButton = getByTestId('pagination-Button3');
    fireEvent.click(nextButton);

    const activePageButton = getByTestId('pagination-Button3');
    expect(activePageButton).toHaveStyle('background-color: hsla(200, 85%, 60%, 0.8)');
  });

  it('handles previous page click', () => {
    const { getByTestId } = render(
      <CarouselContextProvider>
        <Carousel i18n={'Cards'} cards={mockCards} defaultCardsCount={1} defaultActivePage={2}>
          <Card />
        </Carousel>
      </CarouselContextProvider>,
    );
    const previousButton = getByTestId('pagination-Button1');
    fireEvent.click(previousButton);

    const activePageButton = getByTestId('pagination-Button1');
    expect(activePageButton).toHaveStyle('background-color: hsla(200, 85%, 60%, 0.8)');
  });

  it('should call handlePrevPage on ArrowLeft keydown', () => {
    const { getByTestId } = render(
      <CarouselContextProvider>
        <Carousel i18n={'Cards'} cards={mockCards} defaultCardsCount={1} defaultActivePage={3}>
          <Card />
        </Carousel>
      </CarouselContextProvider>,
    );
    const body = getByTestId('carousel-container__body');

    fireEvent.keyDown(body, {
      key: 'ArrowLeft',
      code: 'ArrowLeft',
    });

    const activePageButton = getByTestId('pagination-Button2');
    expect(activePageButton).toHaveStyle('background-color: hsla(200, 85%, 60%, 0.8)');
  });

  it('should call handleNextPage on ArrowRight keydown', () => {
    const { getByTestId } = render(
      <CarouselContextProvider>
        <Carousel i18n={'Cards'} cards={mockCards} defaultCardsCount={1} defaultActivePage={1}>
          <Card />
        </Carousel>
      </CarouselContextProvider>,
    );
    const body = getByTestId('carousel-container__body');

    fireEvent.keyDown(body, {
      key: 'ArrowRight',
      code: 'ArrowRight',
    });

    const activePageButton = getByTestId('pagination-Button2');
    expect(activePageButton).toHaveStyle('background-color: hsla(200, 85%, 60%, 0.8)');
  });

  it('should call onCurrentPage in CustomPaginationBtn', () => {
    const CustomPaginationBtn = (props: any) => {
      const { isActivePage, ...rest } = props;

      return (
        <button {...rest} className={'pagination-Button'} style={{ backgroundColor: isActivePage ? 'orange' : 'grey' }}>
          click
        </button>
      );
    };

    const { container } = render(
      <CarouselContextProvider>
        <Carousel
          i18n={'Cards'}
          cards={mockCards}
          defaultCardsCount={1}
          defaultActivePage={1}
          CustomPaginationBtn={<CustomPaginationBtn />}
        >
          <Card />
        </Carousel>
      </CarouselContextProvider>,
    );
    const btn = container.getElementsByClassName('pagination-Button')[2];

    fireEvent.click(btn);

    expect(btn).toHaveStyle('background-color: orange;');
  });
});
