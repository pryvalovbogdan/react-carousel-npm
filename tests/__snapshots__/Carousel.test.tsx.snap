// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel test render snapshot Carousel 1`] = `
<body>
  <div>
    <div
      class="carousel-container"
      data-testid="carousel-Cards"
    >
      <div
        class="carousel-container__header"
      >
        <div>
          <button
            class="button"
            color="neutral"
            disabled=""
          >
            <a
              class="icon icon__previous"
            />
          </button>
          <button
            class="button"
            color="neutral"
          >
            <a
              class="icon icon__next"
            />
          </button>
        </div>
      </div>
      <div
        class="carousel-container__body"
        data-testid="carousel-container__body"
      >
        <div
          style="height: 0px; width: 100%; max-width: 300px;"
        />
        <div
          style="gap: 1px; display: flex; width: 100%;"
        >
          <div>
            Card 1
          </div>
        </div>
      </div>
      <div
        class="carousel-container__pagination-box"
      >
        <button
          class="pagination-button"
          data-testid="pagination-Button1"
          style="cursor: pointer; background-color: rgba(66, 66, 66, 0.8);"
        />
        <button
          class="pagination-button"
          data-testid="pagination-Button2"
          style="cursor: pointer; background-color: rgba(0, 0, 0, 0.1);"
        />
        <button
          class="pagination-button"
          data-testid="pagination-Button3"
          style="cursor: pointer; background-color: rgba(0, 0, 0, 0.1);"
        />
      </div>
    </div>
  </div>
</body>
`;
